### PR TITLE
spotify-tui: add collection variant patch

### PR DESCRIPTION
**Pull request by MoritzBoehme**
###### Description of changes
Add patch adding the collection variant to rspotify used by spotify-tui.
https://github.com/Rigellute/spotify-tui/issues/960#issuecomment-1108813658

This fixes the issue of getting an error when playing liked songs. https://github.com/Rigellute/spotify-tui/issues/960

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jwijenbergh** [comment]
It fails to build for me:

In this branch:
```bash
nix-build -A spotify-tui .
```

gives:

```bash
error: failed to sync

Caused by:
  failed to load pkg lockfile

Caused by:
  failed to resolve patches for `https://github.com/rust-lang/crates.io-index`

Caused by:
  failed to load source for dependency `rspotify`

Caused by:
  Unable to update /build/source/rspotify-0.10.0

Caused by:
  failed to read `/build/source/rspotify-0.10.0/Cargo.toml`

Caused by:
  No such file or directory (os error 2)

Traceback (most recent call last):
  File "/nix/store/r3nwnppm7n0fbnynfrx2q0dzjz5dgfar-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 42, in <module>
    main()
  File "/nix/store/r3nwnppm7n0fbnynfrx2q0dzjz5dgfar-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 17, in main
    assert list(data.keys()) == ["source"]
AssertionError
error: builder for '/nix/store/a721mj4bc6jzdiw319api0y74fm069mk-spotify-tui-0.25.0-vendor.tar.gz.drv' failed with exit code 1;
       last 10 log lines:
       >   failed to read `/build/source/rspotify-0.10.0/Cargo.toml`
       >
       > Caused by:
       >   No such file or directory (os error 2)
       > Traceback (most recent call last):
       >   File "/nix/store/r3nwnppm7n0fbnynfrx2q0dzjz5dgfar-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 42, in <module>
       >     main()
       >   File "/nix/store/r3nwnppm7n0fbnynfrx2q0dzjz5dgfar-cargo-vendor-normalise/bin/.cargo-vendor-normalise-wrapped", line 17, in main
       >     assert list(data.keys()) == ["source"]
       > AssertionError
       For full logs, run 'nix log /nix/store/a721mj4bc6jzdiw319api0y74fm069mk-spotify-tui-0.25.0-vendor.tar.gz.drv'.
error: 1 dependencies of derivation '/nix/store/3750ndflrd18gxkfrvwa024sxphza6a2-spotify-tui-0.25.0.drv' failed to build
```

Any idea why?

**Comment by MoritzBoehme** [comment]
Oh I also had this error previously. I think because the rspotify-0.10.0 directory does not get linked correctly. But it is weird, that it builds on my machine. I'll clear my cached nix store builds and have another look...

**Comment by MoritzBoehme** [comment]
Maybe you have a better idea how to patch rspotify and use it in the `Cargo.lock`. I haven't really done much with rust yet so I have limited experience.

**Comment by jwijenbergh** [comment]
> Maybe you have a better idea how to patch rspotify and use it in the `Cargo.lock`. I haven't really done much with rust yet so I have limited experience.

I am also honestly not sure how to do this, I don't have too much rust/nix packaging experience. I tried to find similar packages in nixpkgs that do this, but I couldn't find any. Simply replacing a rust crate with a patched crate doesn't seem like an odd thing to do, so I'm not sure why I can't figure it out

**Comment by MoritzBoehme** [comment]
Could you try again now?
Apparently a combination of cargoPatches and patches did the trick...

**Comment by MoritzBoehme** [comment]
Result of `nixpkgs-review pr 170915` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>spotify-tui</li>
  </ul>
</details>

**Comment by jwijenbergh** [comment]
> Could you try again now? Apparently a combination of cargoPatches and patches did the trick...

works now using `nixpkgs-review pr 170915`. I guess cargoPatches is only for the lock file? https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/rust.section.md#building-a-crate-with-an-absent-or-out-of-date-cargolock-file-building-a-crate-with-an-absent-or-out-of-date-cargolock-file

Anyways, I can't seem to reproduce the upstream issue of the inability to play liked song, they play fine for me. But I guess it depends on your spotify account?

If this patch fixes things for you then LGTM, it's better if upstream would move to a newer rspotify version though... But for now this will do.

**Review by jwijenbergh — APPROVED** [review]

**Comment by MoritzBoehme** [comment]
For me, it only occurred when I previously played my liked songs on my mobile phone or via the web client, and never through spotify-tui.
And I agree that an upstream update would be best, but as far as I understood rspotify had a lot of breaking changes, so it would take a considerable effort.

**Comment by jwijenbergh** [comment]
> Yeah I thougt about that as well, but apparently they are just joined together if I read the [source ](https://github.com/NixOS/nixpkgs/blob/fa86dc70893ccc3836b6ac90c70bcde59647724b/pkgs/build-support/rust/build-rust-package/default.nix#L134) correctly. But hey if it works `¯\_(ツ)_/¯`
> 
If it works it works!

> 
> For me, it only occurred when I previously played my liked songs on my mobile phone or via the web client, and never through spotify-tui.
> 

Ah yes, can confirm that this fixes the issue. I could reproduce it by playing something on my phone in liked songs and then opening spotify-tui

> And I agree that an upstream update would be best, but as far as I understood rspotify had a lot of breaking changes, so it would take a considerable effort.

Yeah exactly, rspotify 0.11.0 apparently has a lot of breaking changes/api changes.

Anyways thanks for the patch!

File: pkgs/applications/audio/spotify-tui/default.nix [hunk 1]
```diff
@@ -11,7 +11,40 @@ rustPlatform.buildRustPackage rec {
...
+  patches = [
+    ./Cargo.toml.patch
+  ];
+
+  preBuild = let
+    rspotify = stdenv.mkDerivation rec {
+      pname = "rspotify";
+      version = "0.10.0";
+
+      src = fetchurl {
```
[the remark this answers is not in the document]
Comment (MoritzBoehme) [reply pkgs/applications/audio/spotify-tui/default.nix]: Oh thanks

File: pkgs/applications/audio/spotify-tui/default.nix [hunk 2]
```diff
@@ -11,7 +11,40 @@ rustPlatform.buildRustPackage rec {
...
   };
 
-  cargoSha256 = "sha256-iucI4/iMF+uXRlnMttobu4xo3IQXq7tGiSSN8eCrLM0=";
+  # Use patched rspotify
+  cargoPatches = [
+    ./Cargo.lock.patch
+  ];
+  patches = [
+    ./Cargo.toml.patch
+  ];
```
[the remark this answers is not in the document]
Comment (MoritzBoehme) [reply pkgs/applications/audio/spotify-tui/default.nix:20]: As described above, the derivation strangely won't build if the patches are applied through one patch file.

**Review by jwijenbergh — APPROVED** [review]
Still looks good!

File: pkgs/applications/audio/spotify-tui/default.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwijenbergh) [reply pkgs/applications/audio/spotify-tui/default.nix:20]: As @MoritzBoehme said, we have tried this already

**Pull request merged**